include/cuco/detail/static_multimap.inl

```diff
@@ -0,0 +1,1870 @@
+    if (g.any(equals)) { return true; }
+
+    // we found an empty slot, meaning that the key we're searching for isn't present
+    if (g.any(slot_is_empty)) { return false; }
+
+    // otherwise, all slots in the current window are full with other keys, so we move onto the next
+    // window
+    current_slot = next_slot(current_slot);
+  }
+}
+
+template <typename Key,
+          typename Value,
+          class ProbeSequence,
+          cuda::thread_scope Scope,
+          typename Allocator>
+template <bool uses_vector_load, bool is_outer, typename CG, typename KeyEqual>
+__device__ std::enable_if_t<uses_vector_load, void>
+static_multimap<Key, Value, ProbeSequence, Scope, Allocator>::device_view::count_impl(
+  CG const& g, Key const& k, std::size_t& thread_num_matches, KeyEqual key_equal) noexcept
+{
+  auto current_slot = initial_slot(g, k);
+
+  [[maybe_unused]] bool found_match = false;
+
+  while (true) {
+    value_type arr[2];
+    load_pair_array(&arr[0], current_slot);
+
+    auto const first_slot_is_empty =
+      detail::bitwise_compare(arr[0].first, this->get_empty_key_sentinel());
+    auto const second_slot_is_empty =
+      detail::bitwise_compare(arr[1].first, this->get_empty_key_sentinel());
+    auto const first_equals  = (not first_slot_is_empty and key_equal(arr[0].first, k));
+    auto const second_equals = (not second_slot_is_empty and key_equal(arr[1].first, k));
+
+    if constexpr (is_outer) {
+      if (g.any(first_equals or second_equals)) { found_match = true; }
+    }
+
+    thread_num_matches += (first_equals + second_equals);
+
+    if (g.any(first_slot_is_empty or second_slot_is_empty)) {
+      if constexpr (is_outer) {
+        if ((not found_match) && (g.thread_rank() == 0)) { thread_num_matches++; }
+      }
+      break;
+    }
+
+    current_slot = next_slot(current_slot);
+  }
+}
+
+template <typename Key,
+          typename Value,
+          class ProbeSequence,
+          cuda::thread_scope Scope,
+          typename Allocator>
+template <bool uses_vector_load, bool is_outer, typename CG, typename KeyEqual>
+__device__ std::enable_if_t<not uses_vector_load, void>
+static_multimap<Key, Value, ProbeSequence, Scope, Allocator>::device_view::count_impl(
+  CG const& g, Key const& k, std::size_t& thread_num_matches, KeyEqual key_equal) noexcept
+{
+  auto current_slot = initial_slot(g, k);
+
+  [[maybe_unused]] bool found_match = false;
+
+  while (true) {
+    pair<Key, Value> slot_contents =
+      *reinterpret_cast<cuco::pair_type<Key, Value> const*>(current_slot);
+    auto const& current_key = slot_contents.first;
+
+    auto const slot_is_empty = detail::bitwise_compare(current_key, this->get_empty_key_sentinel());
+    auto const equals        = not slot_is_empty and key_equal(current_key, k);
+
+    if constexpr (is_outer) {
+      if (g.any(equals)) { found_match = true; }
+    }
+
+    thread_num_matches += equals;
```

Comment:
Is it possible that a previous lane contains an empty slot?

For example, suppose we are using linear probing with CG size of 8, and the hash table has 12 slots. The hash table contains the following elements (E represents empty slots).
E E E E 0 1 2 3 4 5 6 7

Suppose we want to count key `0`, and the initial hash to slot 4. 
```
E E E E 0 1 2 3 4 5 6 7
        ^
```
For this iteration, we find a match on lane 0 and none of the lanes have empty slots.

Next iteration, `current_slot` advances to slot 0.
```
E E E E 0 1 2 3 4 5 6 7
^
```
In this iteration, we again find a match on lane 4. The algorithm stops after this iteration since lane 0-3 contains empty slots. As a result, both lane 0 and lane 4 have `thread_num_matches` set to 1, but we should only find one match.

Reply:
Good question!

By invoking `get_valid_capacity` in the multimap constructor, we make sure that the actual capacity will be a prime number times the CG size, and this actual capacity will be no smaller than the requested one. In your example with CG size = 8 and requested capacity =12, the minimal prime number `p` meeting the requirement `p * 8 >= 12` would be `p = 2`. As a result, the actual capacity would be 16 so the CG overlapping situation you described above should not happen.

`p` has to be a prime number in the double hashing case so that the CG window can loop over the whole hash table without missing any slots. As for linear probing, `p` can be any integer meeting the requirement `p * CGsize >= desired_capacity`. The current implementation uses a prime number in both cases for simplicity.

Reply:
Make sense. Thanks for clarifying up!